server/src/main/java/org/candlepin/pki/impl/JSSPKIUtility.java

```diff
@@ -0,0 +1,337 @@
+import java.security.cert.CRLException;
+import java.security.cert.CertificateEncodingException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509CRL;
+import java.security.cert.X509Certificate;
+import java.security.interfaces.RSAPrivateKey;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.RSAPrivateCrtKeySpec;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+/**
+ * PKI utility that uses the JSS crypto provider
+ */
+public class JSSPKIUtility implements PKIProviderUtility {
+    public static final byte[] LINE_SEPARATOR = String.format("%n").getBytes();
+    public static final String SIGNING_ALG_ID = "SHA256withRSA";
+
+    public static final String CRL_PEM_NAME = "X509 CRL";
+    public static final String CERTIFICATE_PEM_NAME = "CERTIFICATE";
+
+    // Note that using RSA PRIVATE KEY instead of PRIVATE KEY will indicate this is
+    // a PKCS1 format instead of a PKCS8.
+    public static final String PRIVATE_KEY_PEM_NAME = "RSA PRIVATE KEY";
+
+    private Configuration config;
+    private SubjectKeyIdentifierWriter subjectKeyWriter;
+    private CertificateReader reader;
+
+    @Inject
+    public JSSPKIUtility(CertificateReader reader, SubjectKeyIdentifierWriter subjectKeyWriter,
+        Configuration config) {
+        this.reader = reader;
+        this.subjectKeyWriter = subjectKeyWriter;
+        this.config = config;
+    }
+
+    @Override
+    public X509Certificate createX509Certificate(String dn, Set<X509ExtensionWrapper> extensions,
+        Set<X509ByteExtensionWrapper> byteExtensions, Date startDate, Date endDate, KeyPair clientKeyPair,
+        BigInteger serialNumber, String alternateName) throws IOException {
+
+        X509CertInfo certInfo = new X509CertInfo();
+        try {
+            X509Certificate caCert = reader.getCACert();
+            byte[] publicKeyEncoded = clientKeyPair.getPublic().getEncoded();
+
+            certInfo.set(X509CertInfo.ISSUER,
+                new CertificateIssuerName(new X500Name(caCert.getSubjectX500Principal().getEncoded())));
+            certInfo.set(X509CertInfo.SERIAL_NUMBER, new CertificateSerialNumber(serialNumber));
+            certInfo.set(X509CertInfo.VALIDITY, new CertificateValidity(startDate, endDate));
+            certInfo.set(X509CertInfo.SUBJECT, new CertificateSubjectName(new X500Name(dn)));
+            certInfo.set(X509CertInfo.KEY,
+                new CertificateX509Key(X509Key.parse(new DerValue(publicKeyEncoded))));
+            certInfo.set(X509CertInfo.ALGORITHM_ID,
+                new CertificateAlgorithmId(AlgorithmId.get(SIGNING_ALG_ID)));
+            certInfo.set(X509CertInfo.VERSION, new CertificateVersion(CertificateVersion.V3));
+
+            X509CertImpl certImpl = new X509CertImpl(certInfo);
+
+            // TODO extensions here
```

Comment:
Is this needed now, or is this for to-be-determined extensions as need arrives in the future?

Reply:
It's for the next sprint basically.